arcgis-ios-sdk-samples/Layers/Browse OGC API feature service/BrowseOGCAPIFeatureServiceViewController.swift

```diff
@@ -0,0 +1,246 @@
+                forName: UITextField.textDidChangeNotification,
+                object: textField,
+                queue: .main
+            ) { [unowned loadAction] _ in
+                let text = textField.text!.trimmingCharacters(in: .whitespacesAndNewlines)
+                // Enable the load button if textfield is not empty.
+                loadAction.isEnabled = !text.isEmpty
+            }
+        }
+        present(alertController, animated: true)
+    }
+    
+    @IBAction func browseLayerInfos(_ sender: UIBarButtonItem) {
+        let selectedIndex = featureCollectionInfos.firstIndex { $0 == selectedInfo }
+        let optionsViewController = OptionsTableViewController(labels: featureCollectionInfos.map(\.title), selectedIndex: selectedIndex) { [self] newIndex in
+            let selectedInfo = featureCollectionInfos[newIndex]
+            let table = AGSOGCFeatureCollectionTable(featureCollectionInfo: selectedInfo)
+            displayLayer(with: table)
+        }
+        optionsViewController.modalPresentationStyle = .popover
+        optionsViewController.preferredContentSize = CGSize(width: 300, height: 300)
+        optionsViewController.presentationController?.delegate = self
+        optionsViewController.popoverPresentationController?.barButtonItem = sender
+        present(optionsViewController, animated: true)
+    }
+    
+    // MARK: UIViewController
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        // Add the source code button item to the right of navigation bar.
+        (navigationItem.rightBarButtonItem as? SourceCodeBarButtonItem)?.filenames = ["BrowseOGCAPIFeatureServiceViewController"]
+    }
+    
+    override func viewDidAppear(_ animated: Bool) {
+        super.viewDidAppear(animated)
+        guard !urlProvided else { return }
+        // Ask user for service URL when the sample has loaded.
+        var completion: ((Result<URL, Error>) -> Void)!
+        completion = { [weak self] result in
+            guard let self = self else { return }
+            switch result {
+            case .success(let serviceURL):
+                self.loadService(url: serviceURL)
+                self.urlProvided = true
+            case .failure(let error):
+                let alertController = UIAlertController(title: "Error", message: error.localizedDescription, preferredStyle: .alert)
+                let tryAgainAction = UIAlertAction(title: "Try Again", style: .default) { _ in
+                    self.askUserForServiceURL(completion: completion)
+                }
+                let cancelAction = UIAlertAction(title: "Cancel", style: .cancel)
+                alertController.addAction(tryAgainAction)
+                alertController.addAction(cancelAction)
+                alertController.preferredAction = tryAgainAction
+                self.present(alertController, animated: true)
+            }
+        }
+        askUserForServiceURL(completion: completion)
```

Comment:
I think I understand this a little bit better now, but I'm still confused by a couple things. Shouldn't the retry mechanism be part of asking the user for the service URL (part of `askUserForServiceURL(completion:)`)? Then the completion handler could just take a `URL?` instead of a result. However, I'm not sure that retrying is even necessary. Instead of allowing the user to enter an invalid URL, when not just make that part of the text validation process? If a instance of `URL` can't be created from the text, don't enable the `Load` button. Would that work?

Reply:
I like the idea of using URL constructor to validate! Will make the change.